metadata_layer.py

```diff
@@ -352,93 +352,492 @@ def __repr__(self):
         )
 
 class FormatData(object):
-    def __init__(self, content_type, drm_scheme, link=None):
+    def __init__(self, content_type, drm_scheme, link=None, rights_uri=None):
         self.content_type = content_type
         self.drm_scheme = drm_scheme
         if link and not isinstance(link, LinkData):
             raise TypeError(
                 "Expected LinkData object, got %s" % type(link)
             )
         self.link = link
+        self.rights_uri = rights_uri
+
+
+
+class MetaToModelUtility(object):
+    """
+    TODO
+    """
+
+    #def mirror_link(self, pool, data_source, link, link_obj, policy):
+    def mirror_link(self, model_object, data_source, link, link_obj, policy):
+        """Retrieve a copy of the given link and make sure it gets
+        mirrored. If it's a full-size image, create a thumbnail and
+        mirror that too.
+
+        The model_object can be either a pool or an edition.
+        """
+
+        if link_obj.rel not in (
+                Hyperlink.IMAGE, Hyperlink.OPEN_ACCESS_DOWNLOAD
+        ):
+            # we only host locally open-source epubs and cover images
+            return
+
+        mirror = policy.mirror
+        http_get = policy.http_get
+
+        _db = Session.object_session(link_obj)
+        original_url = link.href
+        identifier = link_obj.identifier
```

Comment:
We set identifier here and then immediately unset it on line 399. Either we need a sanity check to make sure the link's identifier is the same as the identifier of the model object, or it shouldn't be set here in the first place.